org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/util/SourceFolderHelper.kt

```diff
@@ -106,122 +98,210 @@ class SourceFolderHelper(val project: Project, val sourceFolder: SourceFolder, v
+
+    private abstract class SourceFolderBundle(val tier: ITier<*>, val referenceSourceFolderBundle: SourceFolderBundle?, val sourceFolderHelperInfo: SourceFolderHelperInfo) {
+        private val m_sourceFolderPair = FinalValue<Pair<SourceFolder, IClasspathEntry>?>()
+        private val m_testSourceFolderPair = FinalValue<Pair<SourceFolder, IClasspathEntry>?>()
+        private val m_mainTestSourceFolder = FinalValue<IClasspathEntry?>()
+        private val m_generatedSourceFolder = FinalValue<IClasspathEntry?>()
+
+        protected abstract fun computeSourceFolderPair(): Pair<SourceFolder, IClasspathEntry>?
+
+        fun sourceFolderPair(): Pair<SourceFolder, IClasspathEntry>? = m_sourceFolderPair.computeIfAbsentAndGet { computeSourceFolderPair() }
+
+        fun testSourceFolderPair(): Pair<SourceFolder, IClasspathEntry>? = m_testSourceFolderPair.computeIfAbsentAndGet {
+            findTestSourceFolderPair(sourceFolderHelperInfo.project, tier, sourceFolderPair()?.first, sourceFolderHelperInfo.tierOfModule, sourceFolderHelperInfo.findClasspathEntry)
+        }
+
+        fun mainTestSourceFolder(): IClasspathEntry? = m_mainTestSourceFolder.computeIfAbsentAndGet {
+            findMainSourceFolder(testSourceFolderPair()?.first)?.file?.let { return@computeIfAbsentAndGet sourceFolderHelperInfo.findClasspathEntry(it) }
+        }
+
+        fun generatedSourceFolder(): IClasspathEntry? = m_generatedSourceFolder.computeIfAbsentAndGet {
+            findGeneratedSourceFolder(sourceFolderPair()?.first)?.file?.let { return@computeIfAbsentAndGet sourceFolderHelperInfo.findClasspathEntry(it) }
         }
 
         /**
-         * Finds the closest test pair of [SourceFolder] and [IClasspathEntry] in the given [project] that is a [scoutTier] for the given [sourceFolder].
+         * Finds the closest test pair of [SourceFolder] and [IClasspathEntry] in the given [project] that is a [ITier] for the given [sourceFolder].
          */
-        fun findTestSourceFolderPair(
+        private fun findTestSourceFolderPair(
             project: Project,
-            scoutTier: ScoutTier,
+            tier: ITier<*>,
             sourceFolder: SourceFolder?,
-            scoutTierOfModule: ((Module) -> ScoutTier?)? = null,
+            tierOfModule: ((Module) -> ITier<*>?)? = null,
             findClasspathEntry: (VirtualFile?) -> IClasspathEntry?
         ): Pair<SourceFolder, IClasspathEntry>? {
             if (sourceFolder == null) {
                 return null
             }
-            val myScoutTierOfModule = scoutTierOfModule ?: { S2iScoutTier.valueOf(it) }
-            if (myScoutTierOfModule(sourceFolder.contentEntry.rootModel.module) == scoutTier) {
-                findTestSourceFolderPair(sourceFolder)?.let { findClasspathEntry(it.file) }?.let { return sourceFolder to it }
+            val myTierOfModule = tierOfModule ?: { S2iTier.of(it) }
+            if (myTierOfModule(sourceFolder.contentEntry.rootModel.module) == tier) {
```

Comment:
Use tier.isIncludedIn(myTierOfModule(sf.contentEntry.rootModel.module)) instead of == and centralize with other occurence